### PR TITLE
feat: onboard algolia destination to cdk 2.0

### PR DESCRIPTION
**Pull request by saikumarrs**
## Description of the change

Onboard Algolia destination to CDK 2.0
- Added error handling to router transformation handler and updated Algolia transformer to throw errors on validation failures.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

N/A

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: __tests__/algolia-cdk.test.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        const expected = expectedRouterData[index];
+        try {
+          rtWorkflowEngine = await rtWorkflowEnginePromise;
+          const result = await rtWorkflowEngine.execute(input);
+    
+          // JSONata uses internal implementation for arrays so
+          // they can't be directly compared with others.
+          // So, we need to use serialize and
+          // deserialize to normalize them for comparison.
+          expect(JSON.parse(JSON.stringify(result.output))).toEqual(
```
Comment (koladilip) [inline __tests__/algolia-cdk.test.js:64]: This is fixed in workflow engine now, not required this any more.

File: __tests__/algolia-cdk.test.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        const expected = expectedData[index];
+        try {
+          procWorkflowEngine = await procWorkflowEnginePromise;
+          const result = await procWorkflowEngine.execute(input);
+
+          // JSONata uses internal implementation for arrays so
+          // they can't be directly compared with others.
+          // So, we need to use serialize and
+          // deserialize to normalize them for comparison.
+          expect(JSON.parse(JSON.stringify(result.output))).toEqual(expected);
```
Comment (koladilip) [inline __tests__/algolia-cdk.test.js:33]: This is fixed in workflow engine now, not required this any more.

File: cdk/v2/destinations/algolia/rtWorkflow.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+- name: prepareContext
+  template:
+    $setContext("batchMode", true)
+
+- name: transform
+  externalWorkflow:
+    path: ./procWorkflow.yaml
+  loopOverInput: true
+
+- name: processTransformOutput
```
Comment (koladilip) [inline cdk/v2/destinations/algolia/rtWorkflow.yaml:20]: Can you check latest Pinterest procWorkflow.yaml, simplified these for better readability.

May be we can reuse the template using templatePath feature

File: versionedRouter.js [hunk 4]
```diff
@@ -307,32 +307,54 @@ async function handleValidation(ctx) {
...
-    .filter(
-      resp =>
-        "error" in resp &&
-        _.isObject(resp.statTags) &&
-        !_.isEmpty(resp.statTags)
-    )
-    .forEach(resp => {
-      set(resp, "statTags.errorAt", TRANSFORMER_METRIC.ERROR_AT.RT);
-    });
+  let destType;
```
Comment (koladilip) [inline versionedRouter.js:311]: I am also making changes this same code block so might have to resolve conflicts.

File: versionedRouter.js [hunk 5]
```diff
@@ -307,32 +307,54 @@ async function handleValidation(ctx) {
...
+    if (!routerDestHandler || !routerDestHandler.processRouterDest) {
+      ctx.status = 404;
+      ctx.body = `${destType} doesn't support router transform`;
+      return null;
+    }
+    const allDestEvents = _.groupBy(input, event => event.destination.ID);
+    await Promise.all(
+      Object.entries(allDestEvents).map(async ([destID, desInput]) => {
+        desInput = processDynamicConfig(desInput, "router");
+        const listOutput = await routerDestHandler.processRouterDest(desInput);
```
Comment (koladilip) [inline versionedRouter.js:325]: RT workflow execution is not included here:
I have done that in this: https://github.com/rudderlabs/rudder-transformer/pull/1471

File: versionedRouter.js (hunk 4 above)
Comment (saikumarrs) [reply versionedRouter.js:404]: Addressed.

File: versionedRouter.js (hunk 5 above)
Comment (saikumarrs) [reply versionedRouter.js]: Addressed.

File: __tests__/algolia-cdk.test.js (hunk 1 above)
Comment (saikumarrs) [reply __tests__/algolia-cdk.test.js]: Addressed.

File: __tests__/algolia-cdk.test.js (hunk 2 above)
Comment (saikumarrs) [reply __tests__/algolia-cdk.test.js]: Addressed.

File: cdk/v2/destinations/algolia/rtWorkflow.yaml (hunk 3 above)
Comment (saikumarrs) [reply cdk/v2/destinations/algolia/rtWorkflow.yaml]: For now, I've implemented it independently. Let's refactor it later.

**Review by koladilip — DISMISSED** [review]

**Review by koladilip — APPROVED** [review]

File: versionedRouter.js [hunk 6]
```diff
@@ -412,34 +412,60 @@ async function handleValidation(ctx) {
...
   const respEvents = [];
   let destType;
   try {
     const { input } = ctx.request.body;
     destType = ctx.request.body.destType;
+    if (!Array.isArray(input)) {
+      ctx.status = 400;
+      ctx.body = "input should be an array";
+      return null;
+    }
```
Comment (saikumarrs) [inline versionedRouter.js:442]: Let's revert this change for now, as it doesn't align with the existing non-CDK flow.

Comment (koladilip) [reply versionedRouter.js:442]: Sure

**Review by koladilip — APPROVED** [review]
